framework/CodeInterfaces/RAVEN/RAVENparser.py

```diff
@@ -0,0 +1,295 @@
+      # full path
+      slaveInputFullPath = os.path.abspath(os.path.join(dirName,slaveInput))
+      # check if exists
+      if os.path.exists(slaveInputFullPath):
+        slaveInputBaseDir = os.path.dirname(slaveInput)
+        slaveDir = os.path.join(currentDirName,slaveInputBaseDir.replace(currentDirName,""))
+        if not os.path.exists(slaveDir):
+          os.makedirs(slaveDir)
+        shutil.copy(slaveInputFullPath,slaveDir)
+      else:
+        raise IOError(self.printTag+' ERROR: File "' +slaveInputFullPath+'" has not been found!!!')
+
+
+  def printInput(self,rootToPrint,outfile=None):
+    """
+      Method to print out the new input
+      @ In, rootToPrint, xml.etree.ElementTree.Element, the Element containing the input that needs to be printed out
+      @ In, outfile, string, optional, output file root
+      @ Out, None
+    """
+    xmlObj = xml.dom.minidom.parseString(ET.tostring(rootToPrint))
+    inputAsString = xmlObj.toprettyxml()
+    inputAsString = "".join([s for s in inputAsString.strip().splitlines(True) if s.strip()])
+    if outfile==None:
+      outfile =self.inputfile
+    IOfile = open(outfile,'wb')
+    IOfile.write(inputAsString)
+    IOfile.close()
+
+  def modifyOrAdd(self,modiDictionary={},save=True, allowAdd = False):
+    """
+      modiDictionary a dict of dictionaries of the required addition or modification
+      {"variableToChange":value }
+      @ In, modiDictionary, dict, dictionary of variables to modify
+            syntax:
+            {'Node|SubNode|SubSubNode:value1','Node|SubNode|SubSubNode@attribute:attributeValue|SubSubSubNode':value2
+                      'Node|SubNode|SubSubNode@attribute':value3}
+             TODO: handle added XML nodes
+      @ In, save, bool, optional, True if the original tree needs to be saved
+      @ In, allowAdd, bool, optional, True if the nodes that are not found should be added (additional piece of input)
+      @ Out, returnElement, xml.etree.ElementTree.Element, the tree that got modified
+    """
+    if save:
+      returnElement = copy.deepcopy(self.tree)            #make a copy if save is requested
+    else:
+      returnElement = self.tree                           #otherwise return the original modified
+
+    for node, value in modiDictionary.items():
+      if "|" not in node:
+        raise IOError(self.printTag+' ERROR: the variable '+node.strip()+' does not contain "|" separator and can not be handled!!')
+      changeTheNode = True
+      allowAddNodes, allowAddNodesPath = [], OrderedDict()
+      if "@" in node:
+        # there are attributes that are needed to locate the node
+        splittedComponents = node.split("|")
+        # check the first
+        pathNode = './'
+        attribName = ''
+        for cnt, subNode in enumerate(splittedComponents):
+          splittedComp = subNode.split("@")
+          component = splittedComp[0]
+          attribPath = ""
+          attribConstruct = OrderedDict()
```

Comment:
Just curious, why OrderedDict?  Does the order of the attributes come into play?

Reply:
Yes the orderedDict is needed to detect if we need to change the attribute value or the node itself (see the if statement later on)